iroh-net/src/magicsock/peer_map/endpoint.rs

```diff
@@ -424,38 +456,47 @@ impl Endpoint {
             warn!(
                 "in `DEV_DERP_ONLY` mode, ignoring request to respond to a hole punching attempt."
             );
-            return msgs;
+            return ping_msgs;
         }
 
         self.prune_direct_addresses();
+        warn!(
+            "TODO: direct addrs after pruning: {:#?}",
+            self.direct_addr_state
+        );
 
-        let pings: Vec<_> = self
-            .direct_addr_state
+        self.direct_addr_state
             .iter()
-            .filter_map(|(ep, st)| {
-                if st.needs_ping(&now) {
-                    return Some(*ep);
-                }
-
-                None
+            .filter_map(|(ipp, state)| state.needs_ping(&now).then_some(*ipp))
+            .filter_map(|ipp| {
+                self.start_ping(SendAddr::Udp(ipp.into()), DiscoPingPurpose::Discovery)
             })
-            .collect();
-        let ping_needed = !pings.is_empty();
-        let have_alive_endpoints = self.direct_addr_state.values().any(|e| e.is_alive());
-
-        if !ping_needed {
-            trace!("no ping needed");
-        }
-
-        for ep in pings.into_iter() {
-            if let Some(msg) =
-                self.start_ping(SendAddr::Udp(ep.into()), DiscoPingPurpose::Discovery)
-            {
-                msgs.push(PingAction::SendPing(msg));
-            }
-        }
-
-        if send_call_me_maybe && (ping_needed || !have_alive_endpoints) {
+            .for_each(|msg| ping_msgs.push(PingAction::SendPing(msg)));
+
+        // send_call_me_maybe is:
+        // - true when
+        //   - stayin-alive is sent (a timer on our end)
+        //   - get_send_addrs() is called from poll_send() WHEN no UDP path
+        // - false when
+        //   - handling a **received** call-me-maybe
+
+        // If we were asked for a call-me-maybe we should send it if we don't have a best
+        // addr.  We may not be sending any pings, because we just received a call-me-maybe
+        // and already send pings just now, which remains valid for 5 seconds.
+        //
+        // TODO: double check if there really is any point in still checking beyond
+        // send_call_me_maybe, it posssibly was already checked at the call site that we
+        // don't have a best addr.
+        //
+        // TODO: this could be problematic if we can not punch: we keep sending out
+        // call-me-maybe messages.
+
+        // The if condition here is really trying to be: send call me maybe if we have just
+        // sent some pings, or did send some pings recently.  However maybe we also need to
+        // send it if we don't have a BestADdr yet.
+
+        // if send_call_me_maybe && (have_ping_msgs || !have_alive_endpoints) {
+        if send_call_me_maybe {
```

Comment:
so this only relies on the caller to determine this now?

Reply:
Yes, the logic of whether pings are being sent is wrong.  We need to send a call-me-maybe even if we don't know any UDP addresses to ping.  The alive endpoints (paths really) don't help much either, a path is considered alive too quickly: if it received a ping it is alive.  But an endpoint can only ever be used to send if it received a pong (at which point it would be in best_addr, so that's already checked by the callers as they wouldn't be asking for this if they had one)